File: architecture-skills/object-oriented-programming-paradigm/examples/abstraction/abstraction.php
<?php
/**
 * Abstraction in PHP.
 *
 * Callers see the essential model of a note store — save and load —
 * while each implementation hides its operational detail (an array vs
 * a file on disk). An interface names the contract explicitly.
 */

declare(strict_types=1);

interface NoteStore
{
    public function save(string $key, string $text): void;
    public function load(string $key): ?string;
}

final class MemoryStore implements NoteStore
{
  /** @var array<string, string> */
    private array $notes = [];

    public function save(string $key, string $text): void
    {
        $this->notes[$key] = $text;
    }

    public function load(string $key): ?string
    {
        return $this->notes[$key] ?? null;
    }
}

final class FileStore implements NoteStore
{
    public function __construct(private readonly string $path) {}

    public function save(string $key, string $text): void
    {
        file_put_contents($this->path, "$key\t$text\n", FILE_APPEND);
    }

    public function load(string $key): ?string
    {
        if (!is_file($this->path)) {
            return null;
        }
        $found = null;
        foreach (file($this->path, FILE_IGNORE_NEW_LINES) as $line) {
            [$k, $text] = array_pad(explode("\t", $line, 2), 2, '');
            if ($k === $key) {
                $found = $text;
            }
        }
        return $found;
    }
}

function runSession(NoteStore $store): void
{
    $store->save('groceries', 'eggs, flour');
    $store->save('groceries', 'eggs, flour, milk');
    $store->save('idea', 'hide the how, expose the what');
    echo '  groceries: ', $store->load('groceries'), PHP_EOL;
    echo '  idea:      ', $store->load('idea'), PHP_EOL;
    echo '  missing:   ', var_export($store->load('missing'), true), PHP_EOL;
}

echo "memory-backed store:\n";
runSession(new MemoryStore());

$path = 'abstraction_demo.txt';
try {
    echo "file-backed store:\n";
    runSession(new FileStore($path));
} finally {
    if (is_file($path)) {
        unlink($path);
    }
}

File: architecture-skills/object-oriented-programming-paradigm/examples/abstraction/abstraction.py
"""Abstraction in Python.

Callers see the essential model of a note store -- save and load --
while each implementation hides its operational detail (a dict vs a
file on disk). Python could rely on duck typing alone, but an ABC
names the contract explicitly; the same client runs against both
implementations untouched.
"""

from __future__ import annotations

import os
from abc import ABC, abstractmethod


class NoteStore(ABC):
    """The abstraction: the only surface callers may rely on."""

    @abstractmethod
    def save(self, key: str, text: str) -> None: ...

    @abstractmethod
    def load(self, key: str) -> str | None: ...


class MemoryStore(NoteStore):
    """Operational detail: a dict. Callers never see it."""

    def __init__(self) -> None:
        self._notes: dict[str, str] = {}

    def save(self, key: str, text: str) -> None:
        self._notes[key] = text

    def load(self, key: str) -> str | None:
        return self._notes.get(key)


class FileStore(NoteStore):
    """Operational detail: append-only lines in a file, last write wins."""

    def __init__(self, path: str) -> None:
        self._path = path

    def save(self, key: str, text: str) -> None:
        with open(self._path, "a", encoding="utf-8") as f:
            f.write(f"{key}\t{text}\n")

    def load(self, key: str) -> str | None:
        if not os.path.exists(self._path):
            return None
        found = None
        with open(self._path, encoding="utf-8") as f:
            for line in f:
                k, _, text = line.rstrip("\n").partition("\t")
                if k == key:
                    found = text  # keep scanning: last write wins
        return found


def run_session(store: NoteStore) -> None:
    """Client: written once against the abstraction, never edited when
    the implementation behind it swaps."""
    store.save("groceries", "eggs, flour")
    store.save("groceries", "eggs, flour, milk")  # overwrite
    store.save("idea", "hide the how, expose the what")
    print("  groceries:", store.load("groceries"))
    print("  idea:     ", store.load("idea"))
    print("  missing:  ", store.load("missing"))


if __name__ == "__main__":
    print("memory-backed store:")
    run_session(MemoryStore())

    path = "abstraction_demo.txt"
    try:
        print("file-backed store:")
        run_session(FileStore(path))
    finally:
        if os.path.exists(path):
            os.remove(path)
